src/librustc/infer/nll_relate/mod.rs

```diff
@@ -680,54 +680,58 @@ where
     universe: ty::UniverseIndex,
 }
 
-impl<D> TypeRelation<'me, 'gcx, 'tcx> for TypeGeneralizer<'me, 'gcx, 'tcx, D>
+impl<D> TypeFolder<'gcx, 'tcx> for TypeGeneralizer<'me, 'gcx, 'tcx, D>
 where
     D: TypeRelatingDelegate<'tcx>,
 {
-    fn tcx(&self) -> TyCtxt<'me, 'gcx, 'tcx> {
-        self.infcx.tcx
-    }
+    type Error = TypeError<'tcx>;
 
-    fn tag(&self) -> &'static str {
-        "nll::generalizer"
+    fn tcx(&self) -> TyCtxt<'_, 'gcx, 'tcx> {
+        self.infcx.tcx
     }
 
-    fn a_is_expected(&self) -> bool {
-        true
+    fn use_variances(&self) -> bool {
+        if self.ambient_variance == ty::Variance::Invariant {
```

Comment:
So, in the branches that I was working on, I made the folder track the ambient variance iirc. I don't however remember exactly how I did this =)

Reply:
It's not a thing that Rust handles particularly well